Disable DR after relocate times out randomly

I see this error randomly both locally and in the CI, in the last days we see lot of failures, so this may be a regression in ramen.

```
2024-11-20T19:29:09.997+0200	INFO	appset-deploy-rbd-busybox	dractions/retry.go:115	drpc phase is Relocated
2024-11-20T19:29:09.999+0200	INFO	appset-deploy-rbd-busybox	dractions/retry.go:58	drpc is ready
=== RUN   TestSuites/Exhaustive/appset-deploy-rbd-busybox/Disable
2024-11-20T19:29:09.999+0200	INFO	appset-deploy-rbd-busybox	dractions/actions.go:102	Unprotecting workload
2024-11-20T19:29:09.999+0200	INFO	appset-deploy-rbd-busybox	dractions/actions.go:107	Deleting drpc
    actions_test.go:64: drpc "appset-deploy-rbd-busybox" is not deleted yet before timeout, fail
=== NAME  TestSuites/Exhaustive/appset-deploy-rbd-busybox
    exhaustive_suite_test.go:91: Disable failed
...
--- FAIL: TestSuites (0.05s)
    ...
    --- FAIL: TestSuites/Exhaustive (6.06s)
        ...
        --- FAIL: TestSuites/Exhaustive/appset-deploy-rbd-busybox (1231.36s)
            --- PASS: TestSuites/Exhaustive/appset-deploy-rbd-busybox/Deploy (0.17s)
            --- PASS: TestSuites/Exhaustive/appset-deploy-rbd-busybox/Enable (90.15s)
            --- PASS: TestSuites/Exhaustive/appset-deploy-rbd-busybox/Failover (270.25s)
            --- PASS: TestSuites/Exhaustive/appset-deploy-rbd-busybox/Relocate (270.29s)
            --- FAIL: TestSuites/Exhaustive/appset-deploy-rbd-busybox/Disable (600.49s)
```

Looking at the cluster we see:

On the hub, the drpc is deleting state (good):

```
% kubectl get drpc -A  --context hub                                          
NAMESPACE   NAME                        AGE   PREFERREDCLUSTER   FAILOVERCLUSTER   DESIREDSTATE   CURRENTSTATE
argocd      appset-deploy-rbd-busybox   31m   dr1                dr2               Relocate       Deleting
```

On dr1, the application is running (good), and the vrg/vr were deleted (good):

```
% kubectl get deploy,pod,pvc,vrg,vr -n appset-deploy-rbd-busybox --context dr1
NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/busybox   1/1     1            1           13m

NAME                           READY   STATUS    RESTARTS   AGE
pod/busybox-7d5747dcf9-47b5t   1/1     Running   0          13m

NAME                                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      VOLUMEATTRIBUTESCLASS   AGE
persistentvolumeclaim/busybox-pvc   Bound    pvc-ff343fdd-0a8d-4b2a-90bc-f05e831a11e2   1Gi        RWO            rook-ceph-block   <unset>                 15m
```

On dr2, we see that the application did not complete the cleanup after relocate:

```
% kubectl get deploy,pod,pvc,vrg,vr -n appset-deploy-rbd-busybox --context dr2
NAME                                STATUS        VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      VOLUMEATTRIBUTESCLASS   AGE
persistentvolumeclaim/busybox-pvc   Terminating   pvc-ff343fdd-0a8d-4b2a-90bc-f05e831a11e2   1Gi        RWO            rook-ceph-block   <unset>                 24m

NAME                                                                    DESIREDSTATE   CURRENTSTATE
volumereplicationgroup.ramendr.openshift.io/appset-deploy-rbd-busybox   secondary      Secondary

NAME                                                             AGE   VOLUMEREPLICATIONCLASS   PVCNAME       DESIREDSTATE   CURRENTSTATE
volumereplication.replication.storage.openshift.io/busybox-pvc   24m   vrc-sample               busybox-pvc   secondary      Secondary
```

I think the issue is not waiting for stable state before disabling dr. It may be a bug in ramen that we cannot handle disable dr when relocated did not complete but there is no reason to test this edge case. We need to test the normal case which is:

1. Wait until application is relocated
2. Wait for first replication - meaning that the application is protected on the new cluster
3. Delete the drpc

This is the flow in basic-test, and it works reliably:
https://github.com/RamenDR/ramen/blob/8c376ffb502123d28953bf432923e9b3914ace32/test/basic-test/relocate#L21

In e2e we wait until phase is relocated:
https://github.com/RamenDR/ramen/blob/8c376ffb502123d28953bf432923e9b3914ace32/e2e/dractions/actions.go#L154

This is very early in relocate flow. Relocated did not succeed yet.

Logs from the failing run:
[disable-timeout.gather.tar.gz](https://github.com/user-attachments/files/17834927/disable-timeout.gather.tar.gz)

Failed builds:
- https://github.com/RamenDR/ramen/actions/runs/11940648215/job/33283766375
- https://github.com/RamenDR/ramen/actions/runs/11955746683/job/33336392815
- https://github.com/RamenDR/ramen/actions/runs/11978974524/job/33400129258
- https://github.com/RamenDR/ramen/actions/runs/11996849328/job/33441787711
- https://github.com/RamenDR/ramen/actions/runs/11988822417/job/33424523294
- https://github.com/RamenDR/ramen/actions/runs/11980488460/job/33404837532

DFBUG: https://issues.redhat.com/browse/DFBUGS-969

Improving logging when waiting for readiness #1687 show revealed that we actually wait for both Availabe and PeerReady and for the first replication. The codes was not clear, but adding unneeded call to waitDRPCReady show that we already call it - but disable dr fail randomly.

So this is not a test flake, the test is correct and reveal random failure in ramen in normal disable DR flow.

```
% grep subscr-deploy-rbd-busybox e2e.log
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox
=== PAUSE TestSuites/Exhaustive/subscr-deploy-rbd-busybox
=== CONT  TestSuites/Exhaustive/subscr-deploy-rbd-busybox
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Deploy
2024-12-01T16:59:04.777+0200	INFO	subscr-deploy-rbd-busybox	deployers/subscription.go:37	Deploying workload
2024-12-01T16:59:09.957+0200	INFO	subscr-deploy-rbd-busybox	deployers/retry.go:28	Subscription phase is Propagated
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Enable
2024-12-01T16:59:09.957+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:42	Protecting workload
2024-12-01T16:59:10.048+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:55	Workload running on dr2
2024-12-01T16:59:10.048+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:57	Annotating placement
2024-12-01T16:59:10.058+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:77	Creating drpc
2024-12-01T16:59:10.090+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:00:40.115+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Failover
2024-12-01T17:00:40.115+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:127	Failing over workload
2024-12-01T17:00:40.115+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:00:40.117+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
2024-12-01T17:00:40.122+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:190	Updating drpc failover to dr1
2024-12-01T17:01:10.158+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:97	drpc phase is FailedOver
2024-12-01T17:01:10.158+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:03:10.249+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
2024-12-01T17:03:10.250+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:03:10.254+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Relocate
2024-12-01T17:03:10.254+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:144	Relocating workload
2024-12-01T17:03:10.254+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:03:10.257+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
2024-12-01T17:03:10.264+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:190	Updating drpc relocate to dr2
2024-12-01T17:05:15.371+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:97	drpc phase is Relocated
2024-12-01T17:05:15.372+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:05:15.373+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
2024-12-01T17:05:15.374+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:53	Waiting until drpc is ready
2024-12-01T17:05:15.376+0200	INFO	subscr-deploy-rbd-busybox	dractions/retry.go:65	drpc is ready
=== RUN   TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Disable
2024-12-01T17:05:15.376+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:103	Unprotecting workload
2024-12-01T17:05:15.376+0200	INFO	subscr-deploy-rbd-busybox	dractions/actions.go:110	Deleting drpc
=== NAME  TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Disable
    exhaustive_suite_test.go:87: drpc "subscr-deploy-rbd-busybox" is not deleted yet before timeout, fail
=== NAME  TestSuites/Exhaustive/subscr-deploy-rbd-busybox
        --- FAIL: TestSuites/Exhaustive/subscr-deploy-rbd-busybox (971.23s)
            --- PASS: TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Deploy (5.18s)
            --- PASS: TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Enable (90.20s)
            --- PASS: TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Failover (150.14s)
            --- PASS: TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Relocate (125.12s)
            --- FAIL: TestSuites/Exhaustive/subscr-deploy-rbd-busybox/Disable (600.59s)
```

--- a/e2e/dractions/actions.go
+++ b/e2e/dractions/actions.go
@@ -5,7 +5,6 @@ package dractions
 
 import (
 	"strings"
-	"time"
 
 	ramen "github.com/ramendr/ramen/api/v1alpha1"
 	"github.com/ramendr/ramen/e2e/deployers"
@@ -17,8 +16,6 @@ import (
 
 const (
 	OcmSchedulingDisable = "cluster.open-cluster-management.io/experimental-scheduling-disable"
-
-	FiveSecondsDuration = 5 * time.Second
 )
 
 // If AppSet/Subscription, find Placement

--- a/e2e/dractions/actions.go
+++ b/e2e/dractions/actions.go
@@ -123,7 +123,7 @@ func Failover(ctx types.Context) error {
 
 	log.Info("Failing over workload")
 
-	return failoverRelocate(ctx, ramen.ActionFailover)
+	return failoverRelocate(ctx, ramen.ActionFailover, ramen.FailedOver)
 }
 
 // Determine DRPC
@@ -140,24 +140,19 @@ func Relocate(ctx types.Context) error {
 
 	log.Info("Relocating workload")
 
-	return failoverRelocate(ctx, ramen.ActionRelocate)
+	return failoverRelocate(ctx, ramen.ActionRelocate, ramen.Relocated)
 }
 
-func failoverRelocate(ctx types.Context, action ramen.DRAction) error {
-	name := ctx.Name()
+func failoverRelocate(ctx types.Context, action ramen.DRAction, state ramen.DRState) error {
+	drpcName := ctx.Name()
 	namespace := ctx.Namespace()
-	drpcName := name
 	client := util.Ctx.Hub.Client
 
 	if err := waitAndUpdateDRPC(ctx, client, namespace, drpcName, action); err != nil {
 		return err
 	}
 
-	if action == ramen.ActionFailover {
-		return waitDRPC(ctx, client, namespace, name, ramen.FailedOver)
-	}
-
-	return waitDRPC(ctx, client, namespace, name, ramen.Relocated)
+	return waitDRPC(ctx, client, namespace, drpcName, state)
 }
 
 func waitAndUpdateDRPC(

--- a/e2e/dractions/actions.go
+++ b/e2e/dractions/actions.go
@@ -152,7 +152,11 @@ func failoverRelocate(ctx types.Context, action ramen.DRAction, state ramen.DRSt
 		return err
 	}
 
-	return waitDRPC(ctx, client, namespace, drpcName, state)
+	if err := waitDRPCPhase(ctx, client, namespace, drpcName, state); err != nil {
+		return err
+	}
+
+	return waitDRPCReady(ctx, client, namespace, drpcName)
 }
 
 func waitAndUpdateDRPC(

--- a/e2e/dractions/actionsdiscoveredapps.go
+++ b/e2e/dractions/actionsdiscoveredapps.go
@@ -90,18 +90,18 @@ func FailoverDiscoveredApps(ctx types.Context) error {
 	log := ctx.Logger()
 	log.Info("Failing over workload")
 
-	return failoverRelocateDiscoveredApps(ctx, ramen.ActionFailover)
+	return failoverRelocateDiscoveredApps(ctx, ramen.ActionFailover, ramen.FailedOver)
 }
 
 func RelocateDiscoveredApps(ctx types.Context) error {
 	log := ctx.Logger()
 	log.Info("Relocating workload")
 
-	return failoverRelocateDiscoveredApps(ctx, ramen.ActionRelocate)
+	return failoverRelocateDiscoveredApps(ctx, ramen.ActionRelocate, ramen.Relocated)
 }
 
 // nolint:funlen,cyclop
-func failoverRelocateDiscoveredApps(ctx types.Context, action ramen.DRAction) error {
+func failoverRelocateDiscoveredApps(ctx types.Context, action ramen.DRAction, state ramen.DRState) error {
 	name := ctx.Name()
 	log := ctx.Logger()
 	namespace := ctx.Namespace() // this namespace is in hub
@@ -142,11 +142,7 @@ func failoverRelocateDiscoveredApps(ctx types.Context, action ramen.DRAction) er
 		return err
 	}
 
-	if err = waitDRPCProgression(ctx, client, namespace, name, ramen.ProgressionCompleted); err != nil {
-		return err
-	}
-
-	if err = waitDRPCReady(ctx, client, namespace, name); err != nil {
+	if err = waitDRPC(ctx, client, namespace, name, state); err != nil {
 		return err
 	}
 

--- a/e2e/dractions/actionsdiscoveredapps.go
+++ b/e2e/dractions/actionsdiscoveredapps.go
@@ -142,7 +142,11 @@ func failoverRelocateDiscoveredApps(ctx types.Context, action ramen.DRAction, st
 		return err
 	}
 
-	if err = waitDRPC(ctx, client, namespace, name, state); err != nil {
+	if err := waitDRPCPhase(ctx, client, namespace, name, state); err != nil {
+		return err
+	}
+
+	if err := waitDRPCReady(ctx, client, namespace, name); err != nil {
 		return err
 	}
 

--- a/e2e/dractions/retry.go
+++ b/e2e/dractions/retry.go
@@ -12,6 +12,8 @@ import (
 	"github.com/ramendr/ramen/e2e/types"
 	"github.com/ramendr/ramen/e2e/util"
 	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"open-cluster-management.io/api/cluster/v1beta1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
@@ -48,58 +50,36 @@ func waitDRPCReady(ctx types.Context, client client.Client, namespace string, dr
 	log := ctx.Logger()
 	startTime := time.Now()
 
+	log.Info("Waiting until drpc is ready")
+
 	for {
 		drpc, err := getDRPC(client, namespace, drpcName)
 		if err != nil {
 			return err
 		}
 
-		conditionReady := checkDRPCConditions(drpc)
-		if conditionReady && drpc.Status.LastGroupSyncTime != nil {
+		available := conditionMet(drpc.Status.Conditions, ramen.ConditionAvailable)
+		peerReady := conditionMet(drpc.Status.Conditions, ramen.ConditionPeerReady)
+
+		if available && peerReady && drpc.Status.LastGroupSyncTime != nil {
 			log.Info("drpc is ready")
 
 			return nil
 		}
 
 		if time.Since(startTime) > util.Timeout {
-			if !conditionReady {
-				log.Info("drpc condition 'Available' or 'PeerReady' is not True")
-			}
-
-			if conditionReady && drpc.Status.LastGroupSyncTime == nil {
-				log.Info("drpc LastGroupSyncTime is nil")
-			}
-
-			return fmt.Errorf("drpc %q is not ready yet before timeout, fail", drpcName)
+			return fmt.Errorf("timeout waiting for drpc to become ready (Available: %v, PeerReady: %v, lastGroupSyncTime: %v)",
+				available, peerReady, drpc.Status.LastGroupSyncTime)
 		}
 
 		time.Sleep(util.RetryInterval)
 	}
 }
 
-func checkDRPCConditions(drpc *ramen.DRPlacementControl) bool {
-	available := false
-	peerReady := false
+func conditionMet(conditions []metav1.Condition, conditionType string) bool {
+	condition := meta.FindStatusCondition(conditions, conditionType)
 
-	for _, cond := range drpc.Status.Conditions {
-		if cond.Type == "Available" {
-			if cond.Status != "True" {
-				return false
-			}
-
-			available = true
-		}
-
-		if cond.Type == "PeerReady" {
-			if cond.Status != "True" {
-				return false
-			}
-
-			peerReady = true
-		}
-	}
-
-	return available && peerReady
+	return condition != nil && condition.Status == "True"
 }
 
 func waitDRPCPhase(ctx types.Context, client client.Client, namespace, name string, phase ramen.DRState) error {

--- a/e2e/dractions/retry.go
+++ b/e2e/dractions/retry.go
@@ -141,16 +141,6 @@ func getTargetCluster(client client.Client, namespace, placementName string, drp
 	return targetCluster, nil
 }
 
-// first wait DRPC to have the expected phase, then check DRPC conditions
-func waitDRPC(ctx types.Context, client client.Client, namespace, name string, expectedPhase ramen.DRState) error {
-	// check Phase
-	if err := waitDRPCPhase(ctx, client, namespace, name, expectedPhase); err != nil {
-		return err
-	}
-	// then check Conditions
-	return waitDRPCReady(ctx, client, namespace, name)
-}
-
 func waitDRPCDeleted(ctx types.Context, client client.Client, namespace string, name string) error {
 	log := ctx.Logger()
 	startTime := time.Now()

